server/src/main/java/org/opensearch/rest/SecurityRestFilter.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.rest;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.authn.HttpHeaderToken;
+import org.opensearch.client.node.NodeClient;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static org.opensearch.node.Node.INTERNAL_REALM;
+
+/**
+ * Adds a wrapper to all rest requests to add authentication mechanism
+ *
+ * Reference: <a href="https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/filter/SecurityRestFilter.java">SecurityRestFilter.java</a>
+ */
+public class SecurityRestFilter {
+
+    private static final Logger logger = LogManager.getLogger(SecurityRestFilter.class);
+    public static final String LEGACY_OPENDISTRO_PREFIX = "_opendistro/_security";
+    public static final String PLUGINS_PREFIX = "_plugins/_security";
+    private static final String REGEX_PATH_PREFIX = "/(" + LEGACY_OPENDISTRO_PREFIX + "|" + PLUGINS_PREFIX + ")/" + "(.*)";
+    private static final Pattern PATTERN_PATH_PREFIX = Pattern.compile(REGEX_PATH_PREFIX);
+    private static final String HEALTH_SUFFIX = "health";
+    private static final String WHO_AM_I_SUFFIX = "whoami";
+
+    public SecurityRestFilter() {
+        super();
+        // TODO modify this constructor and add all relevant configs
+    }
+
+    /**
+     * This function adds a wrapper for all incoming rest requests
+     * @param original the original rest handler that will be used to proceed with request once it is authenticated
+     * @return the RestHandler wrapper to original Rest Handler
+     */
+    public RestHandler wrap(RestHandler original) {
+        return (request, channel, client) -> {
+            if (checkAndAuthenticateRequest(request, channel, client)) {
+                // TODO: Do allowlisting and isAdmin check
+                original.handleRequest(request, channel, client);
+            }
+        };
+    }
+
+    /**
+     * Checks if incoming REST request is correctly formed
+     * and authenticates the subject of the request
+     *
+     * @param request the incoming rest request to be authenticated
+     * @param channel the channel to be used to proceed with request handling
+     * @param client the node the that handles the request
+     * @return true if authentication was successful, false otherwise
+     * @throws Exception generic exception rethrown in case authenticate method throws an exception
+     */
+    private boolean checkAndAuthenticateRequest(RestRequest request, RestChannel channel, NodeClient client) throws Exception {
```

Comment:
Consider splitting this into separate formatting and a authenticated checks--may be easier to provide meaningful exception messages.